src/Raven.Server/Documents/TransactionOperationsMerger.cs

```diff
@@ -848,5 +1092,170 @@ public void Dispose()
+                                    txCtx.Transaction = txCtx.Transaction.BeginAsyncCommitAndStartNewTransaction();
+                                    txDisposable = txCtx.Transaction;
+
+                                    previousTx.EndAsyncCommit();
+                                    previousTx.Dispose();
+                                    break;
+                                case TxInstruction.EndAsyncCommit:
+                                    //Todo I think this is not relevant all the async instruction can run as one unit in  FinishAsyncAndStartNew
+                                    //previousTx.EndAsyncCommit();
+                                    break;
+                            }
+                            continue;
+                        }
+
+                        var cmd = ReadCommand(strType, item, context);
+
+                        cmd?.Execute(txCtx, null);
+                    }
+                }
+            }
+        }
+
+        private MergedTransactionCommand ReadCommand(string type, BlittableJsonReaderObject wrapCmdReader, JsonOperationContext context)
+        {
+            if (!wrapCmdReader.TryGet(CommandKey, out BlittableJsonReaderObject commandReader))
+            {
+                return null;
+                //Todo To remove after all MergedTransactionCommand driven class include in switch
+                throw new Exception("Can't read MergedTransactionCommand for replay");
+            }
+            MergedTransactionCommand cmd;
+            switch (type)
+            {
+                case nameof(BatchHandler.MergedBatchCommand):
+                    cmd = BatchHandler.MergedBatchCommand.Deserialize(commandReader, _parent, context);
+                    break;
+                case nameof(DeleteDocumentCommand):
+                    cmd = DeleteDocumentCommand.Deserialize(commandReader, _parent);
+                    break;
+                case nameof(PatchDocumentCommand):
+                    cmd = PatchDocumentCommand.Deserialize(commandReader, _parent, context);
+                    break;
+                case nameof(DatabaseDestination.MergedBatchPutCommand):
+                    cmd = DatabaseDestination.MergedBatchPutCommand.Deserialize(commandReader, _parent, context);
+                    break;
+                case nameof(MergedPutCommand):
+                    cmd = MergedPutCommand.Deserialize(commandReader, _parent, context);
+                    break;
+                case nameof(BulkInsertHandler.MergedInsertBulkCommand):
+                    cmd = BulkInsertHandler.MergedInsertBulkCommand.Deserialize(commandReader, _parent, context, _log);
+                    break;
+                default:
+                    return null;
+            }
+
+            if (cmd == null)
+            {
+                //Todo to decide what to do when can't read command
+                throw new Exception("Can't read MergedTransactionCommand for replay");
```

Comment:
Don't throw `Exception`
If there is an error, using the `PeepingTomStream` to provide context

Reply:
done